# Trabalho Individual 1- Optimização Heurística

# Trabalho realizado por:
- # \textit{ João Francisco Botas, nº104782, CDB1}






### Resolução do Problema recorrendo a programação linear por metas

### Importação das bibliotecas

In [1]:
import pandas as pd
from pulp import LpMaximize, LpMinimize, LpProblem, LpStatus, lpSum, LpVariable
from pulp import GLPK

### Formulação do problema - alínea a)

Encontra-se apresentada no relatório.

### 1º problema com os pesos já estabelecidos - alínea b)

In [2]:
model = LpProblem(name="Programacao_por_metas", sense=LpMinimize)
#
vector=[125,60,55]
peso=[5,10/5,4/5,5]
# Variáveis de decisão
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(1, 3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2, 4)}
#
# Restrições
model += (x[1] <= 6, "toneladas_d1")
model += (x[2] >= 2, "toneladas_d2")
model += (x[3] >= 1, "toneladas_d3")
model += (12 * x[1] + 9 * x[2] + 5 * x[3] + dm[1] >= 125, "alcancar_lucro")
model += (5 * x[1] + 2 * x[2] + 4 * x[3] + dm[2] - dM[2] == 60, "mao_obra")
model += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "travar_investimento")
#
# Função objetivo
obj_func = peso[0]/vector[0] * dm[1] + peso[1]/vector[1] * dm[2] + peso[2]/vector[1] * dM[2] + peso[3]/vector[2] * dM[3]
model += obj_func
# 
model

Programacao_por_metas:
MINIMIZE
0.013333333333333334*dM2 + 0.09090909090909091*dM3 + 0.04*dm1 + 0.03333333333333333*dm2 + 0.0
SUBJECT TO
toneladas_d1: x1 <= 6

toneladas_d2: x2 >= 2

toneladas_d3: x3 >= 1

alcancar_lucro: dm1 + 12 x1 + 9 x2 + 5 x3 >= 125

mao_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 60

travar_investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dM2 Continuous
dM3 Continuous
dm1 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [3]:
status = model.solve(solver=GLPK(msg=True))
#
# Valor óptimo de problema
model.objective.value()
print(f"objective: {model.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão e desvios
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das restrições
#
for name, constraint in model.constraints.items():
    print (f"{name}: {constraint.value()}") 

objective: 1.3359999999999999
x1: 6.0
x2: 3.4
x3: 1.0
dm1: 17.4
dm2: 19.2
dM2: 0.0
dM3: 0.0
toneladas_d1: 0.0
toneladas_d2: 1.4
toneladas_d3: 0.0
alcancar_lucro: -3.552713678800501e-15
mao_obra: 0.0
travar_investimento: 0.0


- Valor objetivo baixo
- Produzidas 6, 3.4 e 1 toneladas dos doces $D_1$,$D_2$ e $D_3$, respetivamente
- Lucro abaixo com desvio de 17.4
- Nº de empregados com menos 19.2 dos 60 atuais

## Resoluções alternativas - alínea c)

### Utilização de Pesos alternativos

#### Solução Alternativa 1

In [4]:
model = LpProblem(name="Sol_alternativa_2", sense=LpMinimize)
#
vector=[125,60,55]
peso=[3,1,1,2]
#
# Variáveis de decisão
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(1, 3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2, 4)}
#
# Restrições
model += (x[1] <= 6, "toneladas_d1")
model += (x[2] >= 2, "toneladas_d2")
model += (x[3] >= 1, "toneladas_d3")
model += (12 * x[1] + 9 * x[2] + 5 * x[3] + dm[1] >= 125, "alcancar_lucro")
model += (5 * x[1] + 2 * x[2] + 4 * x[3] + dm[2] - dM[2] == 60, "mao_obra")
model += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "travar_investimento")
#
# Função objetivo
obj_func = peso[0]/vector[0] * dm[1] + peso[1]/vector[1] * dm[2] + peso[2]/vector[1] * dM[2] + peso[3]/vector[2] * dM[3]
model += obj_func
# 
model

Sol_alternativa_2:
MINIMIZE
0.016666666666666666*dM2 + 0.03636363636363636*dM3 + 0.024*dm1 + 0.016666666666666666*dm2 + 0.0
SUBJECT TO
toneladas_d1: x1 <= 6

toneladas_d2: x2 >= 2

toneladas_d3: x3 >= 1

alcancar_lucro: dm1 + 12 x1 + 9 x2 + 5 x3 >= 125

mao_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 60

travar_investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dM2 Continuous
dM3 Continuous
dm1 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [5]:
status = model.solve(solver=GLPK(msg=True))
#
# Valor óptimo de problema
model.objective.value()
print(f"objective: {model.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão e desvios
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das restrições
#
for name, constraint in model.constraints.items():
    print (f"{name}: {constraint.value()}") 

objective: 0.6070702727272728
x1: 6.0
x2: 5.33333
x3: 1.0
dm1: 0.0
dm2: 15.3333
dM2: 0.0
dM3: 9.66667
toneladas_d1: 0.0
toneladas_d2: 3.33333
toneladas_d3: 0.0
alcancar_lucro: -2.9999999995311555e-05
mao_obra: -4.000000000026205e-05
travar_investimento: -1.9999999999242846e-05


#### Extra Solução Alternativa

Serão tomados valores inversos aos pesos no que toca à mão de obra e ao investimento, dando mais prioridade à mão de obra com peso 2. O peso do lucro continuará igual a 3, pois é o que penso ser mais relevante.

In [6]:
model = LpProblem(name="Sol_alternativa_2", sense=LpMinimize)
#
vector=[125,60,55]
peso=[3,2,2,1]
#
# Variáveis de decisão
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(1, 3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2, 4)}
#
# Restrições
model += (x[1] <= 6, "toneladas_d1")
model += (x[2] >= 2, "toneladas_d2")
model += (x[3] >= 1, "toneladas_d3")
model += (12 * x[1] + 9 * x[2] + 5 * x[3] + dm[1] >= 125, "alcancar_lucro")
model += (5 * x[1] + 2 * x[2] + 4 * x[3] + dm[2] - dM[2] == 60, "mao_obra")
model += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "travar_investimento")
#
# Função objetivo
obj_func = peso[0]/vector[0] * dm[1] + peso[1]/vector[1] * dm[2] + peso[2]/vector[1] * dM[2] + peso[3]/vector[2] * dM[3]
model += obj_func
# 
model

Sol_alternativa_2:
MINIMIZE
0.03333333333333333*dM2 + 0.01818181818181818*dM3 + 0.024*dm1 + 0.03333333333333333*dm2 + 0.0
SUBJECT TO
toneladas_d1: x1 <= 6

toneladas_d2: x2 >= 2

toneladas_d3: x3 >= 1

alcancar_lucro: dm1 + 12 x1 + 9 x2 + 5 x3 >= 125

mao_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 60

travar_investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dM2 Continuous
dM3 Continuous
dm1 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [7]:
status = model.solve(solver=GLPK(msg=True))
#
# Valor óptimo de problema
model.objective.value()
print(f"objective: {model.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão e desvios
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das restrições
#
for name, constraint in model.constraints.items():
    print (f"{name}: {constraint.value()}") 

objective: 0.67972
x1: 6.0
x2: 2.38462
x3: 6.30769
dm1: 0.0
dm2: 0.0
dM2: 0.0
dM3: 37.3846
toneladas_d1: 0.0
toneladas_d2: 0.38461999999999996
toneladas_d3: 5.30769
alcancar_lucro: 2.999999999886427e-05
mao_obra: 0.0
travar_investimento: 1.9999999999242846e-05


### Utilizando MiniMax

In [8]:
model2 = LpProblem(name="Sol_alternativa_2", sense=LpMinimize)
#
# Variáveis de decisão
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(1, 3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2, 4)}
Q= LpVariable("Q", lowBound=0)
#
vector=[125,60,55]
peso=[5,10/5,4/5,5]
# Restrições
model2 += (x[1] <= 6, "toneladas_d1")
model2 += (x[2] >= 2, "toneladas_d2")
model2 += (x[3] >= 1, "toneladas_d3")
model2 += (12 * x[1] + 9 * x[2] + 5 * x[3] + dm[1] >= 125, "alcancar_lucro")
model2 += (5 * x[1] + 2 * x[2] + 4 * x[3] + dm[2] - dM[2] == 60, "mao_obra")
model2 += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "travar_investimento")
model2 += (peso[0]/vector[0] * dm[1] - Q <= 0, "desvio_inf_meta_1" )
model2 += (peso[1]/vector[1] * dm[2] - Q <= 0, "desvio_inf_meta_2" )
model2 += (peso[2]/vector[1] * dM[2] - Q <= 0, "desvio_sup_meta_2" )
model2 += (peso[3]/vector[2] * dM[3] - Q <= 0, "desvio_sup_meta_3" )
#
# Função objetivo em Q
obj_func = Q
model2 += obj_func
#
model2

Sol_alternativa_2:
MINIMIZE
1*Q + 0
SUBJECT TO
toneladas_d1: x1 <= 6

toneladas_d2: x2 >= 2

toneladas_d3: x3 >= 1

alcancar_lucro: dm1 + 12 x1 + 9 x2 + 5 x3 >= 125

mao_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 60

travar_investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

desvio_inf_meta_1: - Q + 0.04 dm1 <= 0

desvio_inf_meta_2: - Q + 0.0333333333333 dm2 <= 0

desvio_sup_meta_2: - Q + 0.0133333333333 dM2 <= 0

desvio_sup_meta_3: - Q + 0.0909090909091 dM3 <= 0

VARIABLES
Q Continuous
dM2 Continuous
dM3 Continuous
dm1 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [9]:
status = model2.solve(solver=GLPK(msg=True))
#
# Valor óptimo de problema
model2.objective.value()
print(f"objective: {model2.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão e desvios
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das restrições
#
for name, constraint in model2.constraints.items():
    print (f"{name}: {constraint.value()}") 

objective: 0.541203
x1: 6.0
x2: 3.42539
x3: 1.72829
dm1: 13.5301
dm2: 16.2361
dM2: 0.0
dM3: 5.95323
toneladas_d1: 0.0
toneladas_d2: 1.4253900000000002
toneladas_d3: 0.7282900000000001
alcancar_lucro: 6.000000000128125e-05
mao_obra: 4.0000000002038405e-05
travar_investimento: 4.0000000002038405e-05
desvio_inf_meta_1: 1.0000000000287557e-06
desvio_inf_meta_2: 3.33333333379926e-07
desvio_sup_meta_2: -0.541203
desvio_sup_meta_3: -2.72727272765394e-07


## Programação por Metas Preemptiva - alínea e)

### 1º nível de prioridade

Nesta solução foram utilizados os pesos do enunciado e todos a 1, mas deixei comentado numa mesma célula porque dão valores semelhantes.

In [10]:
model_premp = LpProblem(name="Programacao_por_metas_preemptiva", sense=LpMinimize)

# pesos iniciais
vector=[125,60,55]
peso=[5,10/5,4/5,5]
# peso=[1,1,1,1]

# Variáveis de decisão
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(2, 3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2, 4)}
#
# Restrições
model_premp += (x[1] <= 6, "toneladas_d1")
model_premp += (x[2] >= 2, "toneladas_d2")
model_premp += (x[3] >= 1, "toneladas_d3")
# model_premp += (12 * x[1] + 9 * x[2] + 5 * x[3] >= 125, "alcancar_lucro")
model_premp += (5 * x[1] + 2 * x[2] + 4 * x[3] + dm[2] - dM[2] == 60, "mao_obra")
model_premp += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "travar_investimento")
#
# Função objetivo
obj_func = peso[1]/vector[1] * dm[2] + peso[2]/vector[1] * dM[2] + peso[3]/vector[2] * dM[3]
model_premp += obj_func
#
model_premp

Programacao_por_metas_preemptiva:
MINIMIZE
0.013333333333333334*dM2 + 0.09090909090909091*dM3 + 0.03333333333333333*dm2 + 0.0
SUBJECT TO
toneladas_d1: x1 <= 6

toneladas_d2: x2 >= 2

toneladas_d3: x3 >= 1

mao_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 60

travar_investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dM2 Continuous
dM3 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [11]:
status = model_premp.solve(solver=GLPK(msg=True))
#
# Valor óptimo de problema
model_premp.objective.value()
print(f"objective: {model_premp.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão e desvios
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das restrições
#
for name, constraint in model_premp.constraints.items():
    print (f"{name}: {constraint.value()}") 

objective: 0.6166666666666667
x1: 6.0
x2: 2.0
x3: 1.875
dm2: 18.5
dM2: 0.0
dM3: 0.0
toneladas_d1: 0.0
toneladas_d2: 0.0
toneladas_d3: 0.875
mao_obra: 0.0
travar_investimento: 0.0


#### Atingir o nível de prioridade 1

Como $d_2^-$=18.5 vemos que o primeiro nível de prioridade não foi atingido e, por isso, só haverá mão de obra para 60-18.5=41,5 -> 41 empregados. Para colmatar isso, iremos suavizar a meta, que é bastante otimista, para esse valor novo de forma a cumprir o primeiro nível de prioridade. 

In [12]:
model_premp = LpProblem(name="Programacao_por_metas_preemptiva", sense=LpMinimize)

# pesos iniciais
vector=[125,60,55]
peso=[5,10/5,4/5,5]
# peso=[1,1,1,1]

# Variáveis de decisão
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(2, 3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2, 4)}
#
# Restrições
model_premp += (x[1] <= 6, "toneladas_d1")
model_premp += (x[2] >= 2, "toneladas_d2")
model_premp += (x[3] >= 1, "toneladas_d3")
#
# Sem a primeira meta
# model_premp += (12 * x[1] + 9 * x[2] + 5 * x[3] >= 125, "alcancar_lucro")
#
model_premp += (5 * x[1] + 2 * x[2] + 4 * x[3] + dm[2] - dM[2] == 41, "mao_obra")
model_premp += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "travar_investimento")
#
# Função objetivo
obj_func = peso[1]/vector[1] * dm[2] + peso[2]/vector[1] * dM[2] + peso[3]/vector[2] * dM[3]
model_premp += obj_func
#
model_premp

Programacao_por_metas_preemptiva:
MINIMIZE
0.013333333333333334*dM2 + 0.09090909090909091*dM3 + 0.03333333333333333*dm2 + 0.0
SUBJECT TO
toneladas_d1: x1 <= 6

toneladas_d2: x2 >= 2

toneladas_d3: x3 >= 1

mao_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 41

travar_investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dM2 Continuous
dM3 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [13]:
status = model_premp.solve(solver=GLPK(msg=True))
#
# Valor óptimo de problema
model_premp.objective.value()
print(f"objective: {model_premp.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão e desvios
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das restrições
#
for name, constraint in model_premp.constraints.items():
    print (f"{name}: {constraint.value()}") 

objective: 0.0
x1: 6.0
x2: 2.0
x3: 1.75
dm2: 0.0
dM2: 0.0
dM3: 0.0
toneladas_d1: 0.0
toneladas_d2: 0.0
toneladas_d3: 0.75
mao_obra: 0.0
travar_investimento: -1.0


### 1º nível de prioridade (com pesos diferentes)

In [14]:
model_premp = LpProblem(name="Programacao_por_metas_preemptiva", sense=LpMinimize)

# pesos iniciais
vector=[125,60,55]
peso=[1,10,10,4]

# Variáveis de decisão
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(2, 3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2, 4)}
#
# Restrições
model_premp += (x[1] <= 6, "toneladas_d1")
model_premp += (x[2] >= 2, "toneladas_d2")
model_premp += (x[3] >= 1, "toneladas_d3")
# model_premp += (12 * x[1] + 9 * x[2] + 5 * x[3] >= 125, "alcancar_lucro")
model_premp += (5 * x[1] + 2 * x[2] + 4 * x[3] + dm[2] - dM[2] == 60, "mao_obra")
model_premp += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "travar_investimento")
#
# Função objetivo
obj_func = peso[1]/vector[1] * dm[2] + peso[2]/vector[1] * dM[2] + peso[3]/vector[2] * dM[3]
model_premp += obj_func
#
model_premp

Programacao_por_metas_preemptiva:
MINIMIZE
0.16666666666666666*dM2 + 0.07272727272727272*dM3 + 0.16666666666666666*dm2 + 0.0
SUBJECT TO
toneladas_d1: x1 <= 6

toneladas_d2: x2 >= 2

toneladas_d3: x3 >= 1

mao_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 60

travar_investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dM2 Continuous
dM3 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [15]:
status = model_premp.solve(solver=GLPK(msg=True))
#
# Valor óptimo de problema
model_premp.objective.value()
print(f"objective: {model_premp.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão e desvios
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das restrições
#
for name, constraint in model_premp.constraints.items():
    print (f"{name}: {constraint.value()}") 

objective: 2.690909090909091
x1: 6.0
x2: 2.0
x3: 6.5
dm2: 0.0
dM2: 0.0
dM3: 37.0
toneladas_d1: 0.0
toneladas_d2: 0.0
toneladas_d3: 5.5
mao_obra: 0.0
travar_investimento: 0.0


#### Atingir o nível de prioridade 1 (com pesos diferentes)

In [16]:
model_premp = LpProblem(name="Programacao_por_metas_preemptiva", sense=LpMinimize)

# pesos iniciais
vector=[125,60,55]
#peso=[5,10/5,4/5,5]
peso=[1,10,10,4]

# Variáveis de decisão
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(2, 3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2, 4)}
#
# Restrições
model_premp += (x[1] <= 6, "toneladas_d1")
model_premp += (x[2] >= 2, "toneladas_d2")
model_premp += (x[3] >= 1, "toneladas_d3")
#
# Sem a primeira meta
# model_premp += (12 * x[1] + 9 * x[2] + 5 * x[3] >= 125, "alcancar_lucro")
#
model_premp += (5 * x[1] + 2 * x[2] + 4 * x[3] + dm[2] - dM[2] == 60, "mao_obra")
model_premp += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55+37, "travar_investimento")
#
# Função objetivo
obj_func = peso[1]/vector[1] * dm[2] + peso[2]/vector[1] * dM[2] + peso[3]/vector[2] * dM[3]
model_premp += obj_func
#
model_premp

Programacao_por_metas_preemptiva:
MINIMIZE
0.16666666666666666*dM2 + 0.07272727272727272*dM3 + 0.16666666666666666*dm2 + 0.0
SUBJECT TO
toneladas_d1: x1 <= 6

toneladas_d2: x2 >= 2

toneladas_d3: x3 >= 1

mao_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 60

travar_investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 92

VARIABLES
dM2 Continuous
dM3 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [17]:
status = model_premp.solve(solver=GLPK(msg=True))
#
# Valor óptimo de problema
model_premp.objective.value()
print(f"objective: {model_premp.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão e desvios
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das restrições
#
for name, constraint in model_premp.constraints.items():
    print (f"{name}: {constraint.value()}") 

objective: 0.0
x1: 6.0
x2: 2.0
x3: 6.5
dm2: 0.0
dM2: 0.0
dM3: 0.0
toneladas_d1: 0.0
toneladas_d2: 0.0
toneladas_d3: 5.5
mao_obra: 0.0
travar_investimento: 0.0


### 2º nível de prioridade

In [18]:
model_premp = LpProblem(name="Programacao_por_metas_preemptiva", sense=LpMinimize)

# pesos iniciais
vector=[125,60,55]
peso=[5,10/5,4/5,5]
#peso=[1,1,1,1]
# Variáveis de decisão
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(1, 3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2, 4)}
#

# Restrições
model_premp += (x[1] <= 6, "toneladas_d1")
model_premp += (x[2] >= 2, "toneladas_d2")
model_premp += (x[3] >= 1, "toneladas_d3")
model_premp += (12 * x[1] + 9 * x[2] + 5 * x[3] + dm[1]>= 125, "alcancar_lucro")
model_premp += (5 * x[1] + 2 * x[2] + 4 * x[3] == 41 , "mao_obra") # 60-18.5= 41 ou 42
model_premp += (5 * x[1] + 5 * x[2] + 8 * x[3] <= 55, "travar_investimento")
#
# Função objetivo
obj_func = peso[0]/vector[0] * dm[1]
model_premp += obj_func
#
model_premp

Programacao_por_metas_preemptiva:
MINIMIZE
0.04*dm1 + 0.0
SUBJECT TO
toneladas_d1: x1 <= 6

toneladas_d2: x2 >= 2

toneladas_d3: x3 >= 1

alcancar_lucro: dm1 + 12 x1 + 9 x2 + 5 x3 >= 125

mao_obra: 5 x1 + 2 x2 + 4 x3 = 41

travar_investimento: 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dm1 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [19]:
status = model_premp.solve(solver=GLPK(msg=True))
#
# Valor óptimo de problema
model_premp.objective.value()
print(f"objective: {model_premp.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão e desvios
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
#for var in dm.values():
    #print(f"{var.name}: {var.value()}")
print(f"{dm[1].name}:{dm[1].value()}")
#for var in dM.values():
    #print(f"{var.name}: {var.value()}")
#
# Valores das restrições
#
for name, constraint in model_premp.constraints.items():
    print (f"{name}: {constraint.value()}") 

objective: 0.79
x1: 6.0
x2: 3.0
x3: 1.25
dm1:19.75
toneladas_d1: 0.0
toneladas_d2: 1.0
toneladas_d3: 0.25
alcancar_lucro: 0.0
mao_obra: 0.0
travar_investimento: 0.0


#### Atingir o nível de prioridade 2

Faremos o mesmo procedimento que no 1º nível de prioridade, mas agora para a meta do lucro que tem desvio $d_1^-=19.75$. Por isso, o lucro está abaixo de 125 mil euros em 19.75 mil euros, ou seja, 125-19.75=105,25 euros.  

In [20]:
model_premp = LpProblem(name="Programacao_por_metas_preemptiva", sense=LpMinimize)

# pesos iniciais
vector=[125,60,55]
peso=[5,10/5,4/5,5]
#peso=[1,1,1,1]
# Variáveis de decisão
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(1, 3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2, 4)}
#
# Restrições
model_premp += (x[1] <= 6, "toneladas_d1")
model_premp += (x[2] >= 2, "toneladas_d2")
model_premp += (x[3] >= 1, "toneladas_d3")
model_premp += (12 * x[1] + 9 * x[2] + 5 * x[3] + dm[1]>= 105.25, "alcancar_lucro")
model_premp += (5 * x[1] + 2 * x[2] + 4 * x[3] + dm[2] - dM[2] == 41 , "mao_obra") # 60-18.5= 41 ou 42
model_premp += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "travar_investimento")
#
# Função objetivo
obj_func = peso[1]/vector[1] * dm[2] + peso[2]/vector[1] * dM[2] + peso[3]/vector[2] * dM[3]
model_premp += obj_func
#
model_premp

Programacao_por_metas_preemptiva:
MINIMIZE
0.013333333333333334*dM2 + 0.09090909090909091*dM3 + 0.03333333333333333*dm2 + 0.0
SUBJECT TO
toneladas_d1: x1 <= 6

toneladas_d2: x2 >= 2

toneladas_d3: x3 >= 1

alcancar_lucro: dm1 + 12 x1 + 9 x2 + 5 x3 >= 105.25

mao_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 41

travar_investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dM2 Continuous
dM3 Continuous
dm1 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [21]:
status = model_premp.solve(solver=GLPK(msg=True))
#
# Valor óptimo de problema
model_premp.objective.value()
print(f"objective: {model_premp.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão e desvios
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das restrições
#
for name, constraint in model_premp.constraints.items():
    print (f"{name}: {constraint.value()}") 

objective: 0.0
x1: 6.0
x2: 3.0
x3: 1.25
dm1: 0.0
dm2: 0.0
dM2: 0.0
dM3: 0.0
toneladas_d1: 0.0
toneladas_d2: 1.0
toneladas_d3: 0.25
alcancar_lucro: 0.0
mao_obra: 0.0
travar_investimento: 0.0
